docs/ref-manual/cvl/expr.md

```diff
@@ -1,85 +1,289 @@
+prover.
 
+If you have rules or invariants using `exists` that are running slowly or
+timing out, you can remove the `exists` by manually computing the value that
+exists.  For example, you might replace
+```cvl
+require (exists uint t . priceAtTime(t) != 0);
+```
+with
 ```cvl
-uint x;
-mathint y = x + 1;
-assert y > x;
+require priceAtTime(startTime) != 0;
 ```
 
-The meaning is the same as in the first snippet since an assignment to a `mathint` variable allows non-overflowing interpretations of the arithmetic operators.
+````
+
+Accessing fields and arrays
+---------------------------
+
+One can access the special fields of built-in types, fields of user-defined
+`struct` types, and members of user-defined `enum` types using the normal
+`expr.field` notation.  Note that as described in {ref}`user-types`,
+access to user-defined types must be qualified by a contract name.
+
+Access to arrays also uses the same syntax as Solidity.
 
-The only case in which arithmetic operators in expressions are allowed to overflow is within arguments passed to functions, or generally, whenever we interact with the code being checked. Solidity functions cannot take values that do not fit within 256 bits. Therefore the tool will report an overflow error if `mathint` variable is passed directly as a function argument.
 
+Contracts, method signatures and their properties
+-------------------------------------------------
+
+Writing the ABI signature for a contract method produces a `method` object,
+which can be used to access the {ref}`method fields <method-type>`.
+
+For example,
 ```cvl
-uint256 MAX_INT = 2^256 - 1;
-foo(MAX_INT + 1); // equivalent to invoking foo(0)
-assert MAX_INT + 1 == 0; // always false, because ‘+’ here is mathematical
-mathint x = MAX_INT + 1;
-foo(x); // error
+method m;
+require m.selector == balanceOf(address).selector
+     || m.selector == transfer(address, uint256).selector;
 ```
+will constrain `m` to be either the `balanceOf` or the `transfer` method.
+
+
+One can determine whether a contract has a particular method using the `s in c`
+where `s` is a method selector and `c` is a contract (either `currentContract`
+or a contract introduced with a {ref}`using statement <using-stmt>`.  For
+example, the statement
+```cvl
+if (burnFrom(address,uint256).selector in currentContract) {
+  ...
+}
+```
+will check that the current contract supports the optional `burnFrom` method.
+
+Special variables and fields
+----------------------------
+
+Several of the CVL types have special fields; see {doc}`types` (particularly
+{ref}`env`, {ref}`method-type`, and {ref}`arrays`).
+
+There are also several built-in variables:
+
+ * `bool lastReverted` and `bool lastHasThrown` are boolean values that indicate whether
+   the most recent contract function reverted or threw an exception.  They are
+   only meaningful if the most recent contract function was called with `@withrevert`
+   (see {ref}`call-expr`).
```

Comment:
maybe explain that they can be used after a call with no revert but will always be false. It's a common mistake:
f@withrevert(e,arg)
gomeGetter() => lastReverted

Reply:
I changed the sentence and added a note about this.  We should consider changing this behavior; I added a CVL design doc